imagekit: Added ImageKit backend

ahnv opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

<!--
Describe the changes here
-->

#### Was the change discussed in an issue or in the forum before?

<!--
Link issues and relevant forum posts here.
-->

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw commented:
This looks like you are getting there.

How are the integration tests proceeding?

Do you need help with anything?

ahnv commented:
> This looks like you are getting there.
> 
> How are the integration tests proceeding?
> 
> Do you need help with anything?

Hey, Thanks a lot for reaching

We were able to fix the unicode issue that i reached out for, all the integration tests are passing
I'll be moving this to done and ready from our side, if you can review it at your end as well

ncw commented:
> We were able to fix the unicode issue that i reached out for, all the integration tests are passing I'll be moving this to done and ready from our side, if you can review it at your end as well

Do you mean the `test_all` integration tests? If so that is excellent - well done :-)

Can you rebase it on master (to fix the conflict) please?

I note that you aren't using `lib/rest`. This means you are missing out on lots of rclone's features like `--dump bodies` and `--user-agent` (I see you've added that in by hand).

At absolute minimum you need to use rclone's transport or http client - see `fs/fshttp` but I prefer the use of `lib/rest` if possible as it is a standard bit of rclone machinery and it makes maintenance a lot easier.

ahnv commented:
> Do you mean the `test_all` integration tests? If so that is excellent - well done :-)

Yes, all those test cases are working

> Can you rebase it on master (to fix the conflict) please?

Sure thing, resolving the conflicts and fixing the lint workflow also

> I note that you aren't using lib/rest. This means you are missing out on lots of rclone's features like --dump bodies and --user-agent (I see you've added that in by hand).
> 
> At absolute minimum you need to use rclone's transport or http client - see fs/fshttp but I prefer the use of lib/rest if possible as it is a standard bit of rclone machinery and it makes maintenance a lot easier.

Sure thing, I'll look into this, i'll get this fixed up as well. 
We'd prefer the user-agent to be not configurable by ours users for internal product adoption metrics.

ahnv commented:
hey @ncw, the PR is ready for review

ahnv commented:
Hey @ncw, following up about the review?
Could we help from our end to get this ready?

ahnv commented:
@ncw rebased with master branch

ncw commented on backend/imagekit/client/media.go:
```diff
@@ -0,0 +1,345 @@
+		Method:  "DELETE",
+		Path:    fmt.Sprintf("/files/%s", fileId),
+		RootURL: ik.Prefix,
+	}, nil, nil)
```
Note that if you supply `nil` as the `response` to `CallJSON` you need to close the body, from the docs

> If response is nil then the resp.Body will be closed only if opts.NoResponse is set.

Can you double check all the calls.

ahnv replied:
Verified and updated

ncw commented on backend/imagekit/client/media.go:
```diff
@@ -0,0 +1,345 @@
+		Method:  "POST",
+		Path:    "/folder",
+		RootURL: ik.Prefix,
+	}, param, nil)
```
`nil` response so body needs closing or set `NoResponse`

ahnv replied:
verified and updated

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+			{
+				Name:     "public_key",
+				Help:     "You can find your ImageKit.io public key in your [dashboard](https://imagekit.io/dashboard/developer/api-keys)",
+				Required: true,
```
This should be marked as `Sensitive: true` maybe?

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+			{
+				Name:     "private_key",
+				Help:     "You can find your ImageKit.io private key in your [dashboard](https://imagekit.io/dashboard/developer/api-keys)",
+				Required: true,
```
Also this?

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+type Fs struct {
+	name     string           // name of remote
+	root     string           // root path
+	opt      *Options         // parsed options
```
We normally embed `Options` here which means `Fs` has its own copy, so `opt      Options`

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+	meta        map[string]string // The object metadata if known - may be nil - with lower case keys
+}
+
+// Check the interfaces are satisfied.
+var (
+	_ fs.Fs           = &Fs{}
```
Can you put this at the end of the file to make it consistent with all the other backends please?

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+func (f *Fs) newObject(ctx context.Context, remote string, file client.File) *Object {
+	remoteFile := strings.TrimLeft(strings.Replace(file.FilePath, f.EncodePath(f.root), "", 1), "/")
+
+	folderPath := f.DecodePath(remoteFile[:strings.LastIndex(remoteFile, "/")+1])
```
Did you mean `path.Dir` by this? If it needs the exact semantics you have, can you put it in a function with a sensible name (`pathDir`?) and a comment and a test please!

    remoteFile[:strings.LastIndex(remoteFile, "/")+1]

There are lots of these below - can you replace them all please!

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+func (f *Fs) NewObject(ctx context.Context, remote string) (fs.Object, error) {
+	r := path.Join(f.root, remote)
+
+	fs.Debugf(r, "Getting object")
```
This debug should go as should any others that would be enountered in normal operation.

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+		FolderName:       folderName,
+	})
+
+	if e != nil {
+		return e
+	}
+
+	return nil
```
Just `return e`? But please call it `err` for consistency.

There are more below which I won't note.

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+	for _, option := range options {
+		switch x := option.(type) {
+		case *fs.RangeOption:
+			offset, count = x.Decode(-1)
```
I'm not sure you can write `-1` here - it should be the size of the file.

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+
+	end := resp.ContentLength
+
+	if partialContent && resp.StatusCode == http.StatusOK {
```
Is returning 200 likely to a `Range` request? This is inefficient fetching stuff the user didn't ask for.

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+// But for unknown-sized objects (indicated by src.Size() == -1), Upload should either
+// return an error or update the object properly (rather than e.g. calling panic).
+func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) error {
+	fs.Debugf(o.remote, "Updating file")
```
Remove debug

There are more below which I won't note.

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+
+	meta["size"] = strconv.FormatUint(o.file.Size, 10)
+
+	meta["fileType"] = string(o.file.FileType)
```
metadata keys are kebab-case by convention, so `file-type`

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+	meta["fileType"] = string(o.file.FileType)
+	meta["height"] = strconv.Itoa(o.file.Height)
+	meta["width"] = strconv.Itoa(o.file.Width)
+	meta["hasAlpha"] = strconv.FormatBool(o.file.HasAlpha)
```
`has-alpha`

There are more below which I won't note

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+		case bool:
+			meta[k] = strconv.FormatBool(v)
+		default:
+			meta[k] = fmt.Sprintf("%v", v)
```
`fmt.Sprint(v)` is more efficient. It would probably do for all the types and make the type switch redundant.

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+
+	metadata = make(fs.Metadata, len(o.meta))
+
+	for k, v := range o.meta {
```
why not keep the metadata in the object as `fs.Metadata`? It would save a conversion.

ncw commented on backend/imagekit/imagekit_test.go:
```diff
@@ -0,0 +1,18 @@
+package imagekit
```
To make additional tests make an `image_internal_test.go` file and put them in there.

ncw commented on backend/imagekit/util.go:
```diff
@@ -0,0 +1,189 @@
+		err = f.pacer.Call(func() (bool, error) {
+			var data *[]client.File
+			var res *http.Response
+			res, data, err = f.ik.Files(ctx, client.FilesOrFolderParam{
```
The pacer call needs to be in the `ik.Files` method.

There are more of these below which I won't comment on.

ahnv replied:
updated

ncw commented on backend/imagekit/util.go:
```diff
@@ -0,0 +1,189 @@
+			}, includeVersions)
+
+			if len(*data) == 0 && len(*data) < 100 {
+				hasMore = false
```
perhaps lose the `hasMore` flag and just use `break` here?

There are more of these below which I won't comment on.

ncw commented on docs/content/imagekit.md:
````diff
@@ -0,0 +1,181 @@
+```
+rclone ls imagekit-media-library:directory
+```
+
````
I don't see the auto generated comment markers - did you insert these by hand?

eg from the box backend

```
{{< rem autogenerated options start" - DO NOT EDIT - instead edit fs.RegInfo in backend/box/box.go then run make backenddocs" >}}
{{< rem autogenerated options stop >}}
```

ahnv replied:
fixed it

ncw commented on fstest/test_all/config.yaml:
```diff
@@ -144,6 +144,9 @@ backends:
    fastlist: false
+ - backend:  "imagekit"
+   remote:   "TestImageKit:"
+   fastlist: true
```
You haven't written a `ListR` method that I could see so this should be `false`

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+	decayConstant = 2
+)
+
+var systemMetadataInfo = map[string]fs.MetadataHelp{
```
I saw other metadata below - it needs to be documented here then it will go in the docs automatically.

Run `bin/make_backend_docs.py imagekit` to update your docs - but make sure you put the autogenerated markers in - see below.

ncw commented on backend/imagekit/imagekit.go:
```diff
@@ -0,0 +1,794 @@
+		return "", fs.ErrorObjectNotFound
+	}
+
+	url, err := f.ik.Url(client.UrlParam{
```
Needs pacer?

ahnv replied:
pacer would not be required for the Url call, it's generated at the client side only, no api call

ncw reviewed (commented):
I've given this a thorough review.

It is moving in the right direction - well done :-) It looks much better using lib/rest and you can use its error parsing which will make life easier (see inline).

I think the pacer calls should probably be in the *ImageKit layer as they are too easy to forget otherwise.

Otherwise see inline.

I'll run the CI too which may find more stuff.

Keep going - we are getting there!

ahnv commented:
> Hi!
> 
> Sorry for the delay!
> 
> Once backends are merged the integration tester runs daily and I need an account for that. Will the free account work for that purpose?
> 
> I'll give the code another review now.

Yes, a free account will work for this purpose easily

ncw commented:
Ping me here when you are ready for another review - thank you :-)

ahnv commented:
Hey @ncw, we had to make some minor changes to the backend for file uploads, everything is in order now.
I have addressed the above comments and rebased the branch with master

ahnv commented on backend/imagekit/client/client.go:
```diff
@@ -0,0 +1,68 @@
+		return nil, fmt.Errorf("ImageKit.io private key is required")
+	}
+
+	client := rest.NewClient(fshttp.NewClient(ctx))
```
The custom handling has been removed

ahnv commented:
Thanks for the catch, we recently made some database level changes that caused the getFileByName call to fail, will have a fix for it soon

ncw commented:
Let me know when you want me to take another look.

ahnv commented:
Everything should be in order now

```
=== RUN   TestIntegration
    fstests.go:423: Using remote "TestImageKit:"
2023/11/22 11:03:44 DEBUG : Creating backend with remote "TestImageKit:rclone-test-vopotuf3wenotaw7nazevig8"
2023/11/22 11:03:44 DEBUG : Using config file from "/Users/spyder/.config/rclone/rclone.conf"
=== RUN   TestIntegration/FsCheckWrap
    fstests.go:460: Skipping FsCheckWrap on this Fs
=== RUN   TestIntegration/FsCommand
    fstests.go:492: No commands in this remote
=== RUN   TestIntegration/FsRmdirNotFound
=== RUN   TestIntegration/FsString
=== RUN   TestIntegration/FsName
=== RUN   TestIntegration/FsRoot
=== RUN   TestIntegration/FsRmdirEmpty
=== RUN   TestIntegration/FsMkdir
=== RUN   TestIntegration/FsMkdir/FsMkdirRmdirSubdir
2023/11/22 11:03:49 DEBUG : dir/subdir: Making directory
2023/11/22 11:03:52 INFO  : dir/subdir: Removing directory
2023/11/22 11:03:55 INFO  : dir: Removing directory
=== RUN   TestIntegration/FsMkdir/FsListEmpty
=== RUN   TestIntegration/FsMkdir/FsListDirEmpty
=== RUN   TestIntegration/FsMkdir/FsListRDirEmpty
    fstests.go:384: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsListDirNotFound
2023/11/22 11:03:59 ERROR : does not exist: error listing: directory not found
=== RUN   TestIntegration/FsMkdir/FsListRDirNotFound
    fstests.go:384: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsEncoding
=== RUN   TestIntegration/FsMkdir/FsEncoding/control_chars
    fstests.go:689: testing "␀␁␂␃␄␅␆␇␈␉␊␋␌␍␎␏␐␑␒␓␔␕␖␗␘␙␚␛␜␝␞␟␡"
=== RUN   TestIntegration/FsMkdir/FsEncoding/dot
    fstests.go:689: testing "．"
=== RUN   TestIntegration/FsMkdir/FsEncoding/dot_dot
    fstests.go:689: testing "．．"
=== RUN   TestIntegration/FsMkdir/FsEncoding/punctuation
    fstests.go:689: testing "!\"#$%&'()*+,-.／:;<=>?@[\\]^_`{|}~"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_space
    fstests.go:689: testing " leading space"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_tilde
    fstests.go:689: testing "~leading tilde"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_CR
    fstests.go:689: testing "␍leading CR"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_LF
    fstests.go:689: testing "␊leading LF"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_HT
    fstests.go:689: testing "␉leading HT"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_VT
    fstests.go:689: testing "␋leading VT"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_dot
    fstests.go:689: testing ".leading dot"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_space
    fstests.go:689: testing "trailing space "
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_CR
    fstests.go:689: testing "trailing CR␍"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_LF
    fstests.go:689: testing "trailing LF␊"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_HT
    fstests.go:689: testing "trailing HT␉"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_VT
    fstests.go:689: testing "trailing VT␋"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_dot
    fstests.go:689: testing "trailing dot."
=== RUN   TestIntegration/FsMkdir/FsEncoding/invalid_UTF-8
    fstests.go:689: testing "invalid utf-8\xfe"
=== RUN   TestIntegration/FsMkdir/FsEncoding/URL_encoding
    fstests.go:689: testing "test%46.txt"
=== RUN   TestIntegration/FsMkdir/FsNewObjectNotFound
=== RUN   TestIntegration/FsMkdir/FsPutError
=== RUN   TestIntegration/FsMkdir/FsPutZeroLength
=== RUN   TestIntegration/FsMkdir/FsOpenWriterAt
    fstests.go:762: FS has no OpenWriterAt interface
=== RUN   TestIntegration/FsMkdir/FsOpenChunkWriter
    fstests.go:795: FS has no OpenChunkWriter interface
=== RUN   TestIntegration/FsMkdir/FsChangeNotify
    fstests.go:844: FS has no ChangeNotify interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListDirFile2
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRDirFile2
    fstests.go:384: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListR
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRSubdir
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListDirRoot
2023/11/22 11:07:07 DEBUG : Creating backend with remote "TestImageKit:"
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRDirRoot
    fstests.go:384: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListSubdir
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRSubdir#01
    fstests.go:384: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListLevel2
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRLevel2
    fstests.go:384: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListFile1
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsNewObject
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsNewObjectCaseInsensitive
    fstests.go:1097: Not Case Insensitive
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListFile1and2
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsNewObjectDir
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsPurge
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsPurgeRoot
2023/11/22 11:07:41 DEBUG : Creating backend with remote "TestImageKit:rclone-test-vopotuf3wenotaw7nazevig8/dirToPurgeFromRoot"
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsCopy
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsMove
    fstests.go:1242: FS has no Mover interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsDirMove
    fstests.go:1313: FS has no DirMover interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsRmdirFull
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsPrecision
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectString
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectFs
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectRemote
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectHashes
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectModTime
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectMimeType
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectMetadata
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectSetModTime
    fstests.go:1542: can't set modified time
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectSize
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectOpen
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectOpenSeek
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectPartialRead
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectUpdate
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectStorable
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsIsFile
2023/11/22 11:08:08 DEBUG : Creating backend with remote "TestImageKit:rclone-test-vopotuf3wenotaw7nazevig8/hello? sausage/êé/Hello, 世界/ \" ' @ < > & ? + ≠/z.txt"
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsIsFileNotFound
2023/11/22 11:08:09 DEBUG : Creating backend with remote "TestImageKit:rclone-test-vopotuf3wenotaw7nazevig8/not found.txt"
2023/11/22 11:08:09 ERROR : : error listing: directory not found
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FromRoot
    fstests.go:1684: Opening root remote "TestImageKit:" path "rclone-test-vopotuf3wenotaw7nazevig8" from "TestImageKit:rclone-test-vopotuf3wenotaw7nazevig8"
2023/11/22 11:08:09 DEBUG : Creating backend with remote "TestImageKit:"
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FromRoot/List
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FromRoot/ListEntries
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FromRoot/ListR
    fstests.go:1718: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FromRoot/Put
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FromRoot/Put/Remove
=== RUN   TestIntegration/FsMkdir/FsPutFiles/PublicLink
    fstests.go:1825: skipping directory tests as not supported on this backend
=== RUN   TestIntegration/FsMkdir/FsPutFiles/SetTier
    fstests.go:397: FS has no SetTier & GetTier interfaces
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectCheckWrap
    fstests.go:1882: Not a wrapping Fs
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectRemove
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectAbout
    fstests.go:1916: FS does not support About
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsPutStream
    fstests.go:1934: FS has no PutStream interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/Internal
    fstests.go:1978: *imagekit.Fs does not implement InternalTester
=== RUN   TestIntegration/FsMkdir/FsPutChunked
    fstests.go:1999: *imagekit.Fs does not implement SetUploadChunkSizer
=== RUN   TestIntegration/FsMkdir/FsUploadUnknownSize
=== RUN   TestIntegration/FsMkdir/FsUploadUnknownSize/FsPutUnknownSize
=== RUN   TestIntegration/FsMkdir/FsUploadUnknownSize/FsUpdateUnknownSize
=== RUN   TestIntegration/FsMkdir/FsRootCollapse
2023/11/22 11:08:45 DEBUG : Creating backend with remote "TestImageKit:rclone-test-vopotuf3wenotaw7nazevig8/deeper/nonexisting/directory"
2023/11/22 11:08:47 ERROR : : error listing: directory not found
=== RUN   TestIntegration/FsShutdown
    fstests.go:2195: Shutdown method not supported
--- PASS: TestIntegration (303.19s)
    --- SKIP: TestIntegration/FsCheckWrap (0.00s)
    --- SKIP: TestIntegration/FsCommand (0.00s)
    --- PASS: TestIntegration/FsRmdirNotFound (0.29s)
    --- PASS: TestIntegration/FsString (0.00s)
    --- PASS: TestIntegration/FsName (0.00s)
    --- PASS: TestIntegration/FsRoot (0.00s)
    --- PASS: TestIntegration/FsRmdirEmpty (1.13s)
    --- PASS: TestIntegration/FsMkdir (299.52s)
        --- PASS: TestIntegration/FsMkdir/FsMkdirRmdirSubdir (7.79s)
        --- PASS: TestIntegration/FsMkdir/FsListEmpty (0.85s)
        --- PASS: TestIntegration/FsMkdir/FsListDirEmpty (0.85s)
        --- SKIP: TestIntegration/FsMkdir/FsListRDirEmpty (0.00s)
        --- PASS: TestIntegration/FsMkdir/FsListDirNotFound (0.28s)
        --- SKIP: TestIntegration/FsMkdir/FsListRDirNotFound (0.00s)
        --- PASS: TestIntegration/FsMkdir/FsEncoding (158.87s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/control_chars (8.57s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/dot (7.88s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/dot_dot (8.02s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/punctuation (8.26s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_space (8.47s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_tilde (7.84s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_CR (7.78s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_LF (7.88s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_HT (8.01s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_VT (8.68s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_dot (8.01s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_space (8.66s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_CR (9.60s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_LF (9.26s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_HT (8.85s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_VT (8.43s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_dot (8.34s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/invalid_UTF-8 (7.74s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/URL_encoding (7.73s)
        --- PASS: TestIntegration/FsMkdir/FsNewObjectNotFound (1.15s)
        --- PASS: TestIntegration/FsMkdir/FsPutError (0.58s)
        --- PASS: TestIntegration/FsMkdir/FsPutZeroLength (4.11s)
        --- SKIP: TestIntegration/FsMkdir/FsOpenWriterAt (0.00s)
        --- SKIP: TestIntegration/FsMkdir/FsOpenChunkWriter (0.00s)
        --- SKIP: TestIntegration/FsMkdir/FsChangeNotify (0.00s)
        --- PASS: TestIntegration/FsMkdir/FsPutFiles (110.80s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListDirFile2 (4.25s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsListRDirFile2 (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListR (4.33s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListRSubdir (3.43s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListDirRoot (0.62s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsListRDirRoot (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListSubdir (1.71s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsListRSubdir#01 (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListLevel2 (1.71s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsListRLevel2 (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListFile1 (4.27s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsNewObject (0.57s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsNewObjectCaseInsensitive (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListFile1and2 (4.50s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsNewObjectDir (0.28s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsPurge (12.47s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsPurgeRoot (12.87s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsCopy (8.33s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsMove (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsDirMove (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsRmdirFull (0.85s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsPrecision (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectString (0.56s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectFs (0.56s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectRemote (0.56s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectHashes (0.57s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectModTime (0.56s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectMimeType (0.57s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectMetadata (0.56s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectSetModTime (0.57s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectSize (0.56s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectOpen (0.62s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectOpenSeek (0.60s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange (0.59s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectPartialRead (0.57s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectUpdate (4.47s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectStorable (0.56s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsIsFile (1.14s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsIsFileNotFound (0.57s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FromRoot (18.19s)
                --- PASS: TestIntegration/FsMkdir/FsPutFiles/FromRoot/List (4.49s)
                --- PASS: TestIntegration/FsMkdir/FsPutFiles/FromRoot/ListEntries (0.86s)
                --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FromRoot/ListR (0.00s)
                --- PASS: TestIntegration/FsMkdir/FsPutFiles/FromRoot/Put (12.84s)
                    --- PASS: TestIntegration/FsMkdir/FsPutFiles/FromRoot/Put/Remove (4.56s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/PublicLink (1.41s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/SetTier (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/ObjectCheckWrap (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectRemove (5.12s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/ObjectAbout (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsPutStream (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/Internal (0.00s)
        --- SKIP: TestIntegration/FsMkdir/FsPutChunked (0.00s)
        --- PASS: TestIntegration/FsMkdir/FsUploadUnknownSize (11.13s)
            --- PASS: TestIntegration/FsMkdir/FsUploadUnknownSize/FsPutUnknownSize (3.88s)
            --- PASS: TestIntegration/FsMkdir/FsUploadUnknownSize/FsUpdateUnknownSize (7.25s)
        --- PASS: TestIntegration/FsMkdir/FsRootCollapse (0.84s)
    --- SKIP: TestIntegration/FsShutdown (0.00s)
PASS
ok      github.com/rclone/rclone/backend/imagekit       303.606s
```

ahnv commented:
Hey @ncw, i see that the MR was closed, was this intentional?

ncw commented:
Very sorry was a typo in a git commit. Will re-open.

ncw reviewed (approved):
I've given this a last once over and checked all the integration tests are good, so I'm going to merge this now :-)

Thank you!

Please send any followups as new Pull requests.

This will go into 1.65 which should be out in a couple of days.

ahnv commented:
Thanks a lot
